src/site/content/en/blog/javascript-this/index.md

```diff
@@ -0,0 +1,362 @@
+authors:
+  - jakearchibald
+date: 2021-02-23
+# updated: 2021-02-19
+hero: image/CZmpGM8Eo1dFe0KNhEO9SGO8Ok23/cePCOGeXNFT6WCy85gb4.png
+alt: this 🤔
+tags:
+  - blog # blog is a required tag for the article to show up in the blog.
+  - javascript
+---
+
+JavaScript's `this` is the butt of many jokes, and that's because, well, it's pretty complicated. However, I've seen developers do much-more-complicated and domain-specific things to avoid dealing with this `this`. If you're unsure about `this`, hopefully this will help. This is my `this` guide.
+
+I'm going to start with the most specific situation, and end with the least-specific. This article is kinda like a big `if (…) … else if () … else if (…) …`, so you can go straight to the first section that matches the code you're looking at.
+
+1. [If the function is defined as an arrow function](#arrow-functions)
+1. [Otherwise, if the function/class is called with `new`](#new)
+1. [Otherwise, if the function has a 'bound' `this` value](#bound)
+1. [Otherwise, if `this` is set at call-time](#call-apply)
+1. [Otherwise, if the function is called via a parent object (`parent.func()`)](#object-member)
+1. [Otherwise, if the function or parent scope is in strict mode](#strict)
+1. [Otherwise](#otherwise)
+
+## If the function is defined as an arrow function: {: #arrow-functions }
+
+```js
+const arrowFunction = () => {
+  console.log(this);
+};
+```
+
+In this case, the value of `this` is _always_ the same as `this` in the parent scope:
+
+```js
+const outerThis = this;
+
+const arrowFunction = () => {
+  // Always logs `true`:
+  console.log(this === outerThis);
+};
+```
+
+Arrow functions are great because the inner value of `this` can't be changed, it's _always_ the same as the outer `this`.
+
+### Other examples
+
+With arrow functions, the value of `this` _can't_ be changed with [`bind`](#bound):
+
+```js
+// Logs `true` - bound `this` value is ignored:
+arrowFunction.bind({foo: 'bar'})();
+```
+
+With arrow functions, the value of `this` _can't_ be changed with [`call` or `apply`](#call-apply):
+
+```js
+// Logs `true` - called `this` value is ignored:
+arrowFunction.call({foo: 'bar'});
+// Logs `true` - applied `this` value is ignored:
+arrowFunction.apply({foo: 'bar'});
+```
+
+With arrow functions, the value of `this` _can't_ be changed by calling the function as a member of another object:
+
+```js
+const obj = {arrowFunction};
+// Logs `true` - parent object is ignored:
+obj.arrowFunction();
+```
+
+With arrow functions, the value of `this` _can't_ be changed by calling the function as a constructor:
+
+```js
+// TypeError: arrowFunction is not a constructor
+new arrowFunction();
+```
+
+### 'Bound' instance methods
+
+With instance methods, if you want to ensure `this` always refers to the class instance, the best way is to use arrow functions and [class fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Public_class_fields):
```

Comment:
Curious. You write "the best way", which suggests not the only way. What other alternatives exist? That being said, I love this pattern!

Reply:
I see this pattern a lot:

```js
class Whatever {
  someMethod() {
    // …
  }
  constructor() {
    this.someMethod = this.someMethod.bind(this);
  }
}
```

And:

```js
class Whatever {
  constructor() {
    this.someMethod = () => {
      // …
    };
  }
}
```